raiden-ts/src/transfers/epics/withdraw.ts

```diff
@@ -1,78 +1,370 @@
+          assert(channel?.state === ChannelState.open, 'channel not open');
+          assert(action.meta.totalWithdraw.gt(channel.own.withdraw), 'withdraw already performed');
+          const req = channel.own.pendingWithdraws.find(
+            (req) =>
+              req.type === MessageType.WITHDRAW_REQUEST &&
+              req.total_withdraw.eq(action.payload.message.total_withdraw) &&
+              req.expiration.eq(action.payload.message.expiration),
+          );
+          assert(req, 'no matching WithdrawRequest found');
+          const { tokenNetwork } = action.meta;
+          const { signer: onchainSigner } = chooseOnchainAccount(
+            { signer, address, main },
+            configSubkey,
+          );
+          const tokenNetworkContract = getContractWithSigner(
+            getTokenNetworkContract(tokenNetwork),
+            onchainSigner,
+          );
+
+          return tokenNetworkContract.functions.setTotalWithdraw(
+            channel.id,
+            address,
+            action.meta.totalWithdraw,
+            action.meta.expiration,
+            req.signature,
+            action.payload.message.signature,
+          );
+        }),
+        assertTx('setTotalWithdraw', ErrorCodes.WITHDRAW_TRANSACTION_FAILED, { log }),
+        map((receipt) =>
+          withdraw.success(
+            {
+              txHash: receipt.transactionHash,
+              txBlock: receipt.blockNumber,
+              confirmed: undefined,
+            },
+            action.meta,
+          ),
+        ),
+        catchError((err) => of(withdraw.failure(err, action.meta))),
       ),
     ),
   );
 
-const confMult = [
-  ['partner', 1],
-  ['own', 2],
-] as readonly ['own' | 'partner', number][];
+/**
+ * Upon WithdrawConfirmation, send partner the respective Processed message
+ *
+ * SDK-based clients (with caps.noDelivery set) don't need it, so skip
+ *
+ * @param action$ - Observable of newBlock actions
+ * @param state$ - Observable of RaidenStates
+ * @param deps - Epics dependencies
+ * @param deps.signer - Signer instance
+ * @param deps.log - Logger instance
+ * @param deps.latest$ - Latest observable
+ * @returns Observable of withdrawExpired actions
+ */
+export const withdrawConfirmationProcessedEpic = (
+  action$: Observable<RaidenAction>,
+  {}: Observable<RaidenState>,
+  { signer, log, latest$ }: RaidenEpicDeps,
+): Observable<messageSend.request> => {
+  const cache = new LruCache<string, Signed<Processed>>(32);
+  return action$.pipe(
+    filter(withdrawMessage.success.is),
+    filter((action) => action.meta.direction === Direction.SENT),
+    concatMap((action) =>
+      latest$.pipe(
+        first(),
+        filter(
+          ({ presences }) =>
+            // if caps.noDelivery is set for partner, they don't need this Processed either
+            action.meta.partner in presences &&
+            !presences[action.meta.partner].payload.caps?.[Capabilities.NO_DELIVERY],
+        ),
```

Comment:
Common pattern?

Reply:
I think there're just one other place which checks for `noDelivery` capability, not sure if it's worth to split, but yes, we could look into it in a future refactor. Thanks